joystreamjs/content/src/creatorPayout.ts

```diff
@@ -0,0 +1,261 @@
+  const payloadLengthInBytes = Long.fromNumber(serializedPayloadWithEmptyHeaderFields.byteLength)
+
+  for (let i = 0; i < numberOfChannels; i++) {
+    const channelPayoutRecord = CreatorPayoutPayload.Body.CreatorPayout.encode(body.creatorPayouts![i]).finish()
+    const indexOfChannelPayoutRecord = serializedPayloadWithEmptyHeaderFields.indexOf(channelPayoutRecord)
+    // set correct byteOffsets
+    payload.header.creatorPayoutByteOffsets![i].byteOffset = Long.fromNumber(
+      indexOfChannelPayoutRecord - lengthOfVarintEncodedMessageSize(Buffer.from(channelPayoutRecord).byteLength)
+    )
+  }
+
+  // set correct payload length
+  payload.header.payloadLengthInBytes = payloadLengthInBytes
+  // serialize payload again
+  const serializedPayload = CreatorPayoutPayload.encode(payload).finish()
+  return serializedPayload
+}
+
+/**
+ * Generate merkle root from the serialized payload
+ * @param inputFilePath path to protobuf serialized payload file
+ * @returns merkle root of the cashout vector
+ */
+export async function generateMerkleRoot(inputFilePath: string): Promise<string> {
+  const serializedHeader = await serializedPayloadHeader(inputFilePath)
+  const header = CreatorPayoutPayload.Header.decode(serializedHeader)
+
+  // Any payout record can be used to generate the merkle root,
+  // here first record from creator payouts payload is used
+  const recordByteOffset = header.creatorPayoutByteOffsets.shift()!.byteOffset.toNumber()
+  const record = await creatorPayoutRecordAtByteOffset(inputFilePath, recordByteOffset)
+  return generateMerkleRootFromCreatorPayout(record)
+}
+
+/**
+ * Generate merkle root from branch of creator payout record
+ * @param creatorPayout create payout record
+ * @returns merkle root of the cashout vector
+ */
+export function generateMerkleRootFromCreatorPayout(creatorPayout: CreatorPayoutPayload.Body.CreatorPayout): string {
+  // item = c_i||p_i||m_i
+  const item =
+    creatorPayout.channelId.toString() +
+    creatorPayout.cumulativePayoutOwed.toString() +
+    blake2AsHex(creatorPayout.payoutRationale)
+
+  const merkleRoot = creatorPayout.merkleBranch.reduce((hashV, el) => {
+    if (el.side === CreatorPayoutPayload.Body.CreatorPayout.Side.Right) {
+      return blake2AsHex(hashV + el.hash)
+    } else return blake2AsHex(el.hash + hashV)
+  }, blake2AsHex(item))
+
+  return merkleRoot
+}
+
+/**
+ * fetch creator payouts payload from remote source
+ * @param context `Header | Body` based on part to be fetch from remote source
+ * @param offset If `context` arg is `Body` then offset of payload record must be provided
+ **/
+async function fetchCreatorPayout(
+  context: PayloadContext,
+  offset?: number
+): Promise<CreatorPayoutPayload.Header | CreatorPayoutPayload.Body.CreatorPayout> {
+  // HTTP Url of remote host where payload is stored
+  let url = process.env.CREATOR_PAYOUT_DATA_URL as string
```

Comment:
In the browser env, we won't be able to define this at build time since this URL can be dynamic. So it need to provided via some config/parameter instead

Reply:
Just making a note, not a priority at all from my perspective, realistically, the earliest time for using this in atlas is like 2 months